dll/shellext/zipfldr/CZipExtract.cpp

```diff
@@ -103,20 +150,74 @@ class CZipExtract :
             if (m_pExtract->m_DirectoryChanged)
                 UpdateDirectory();
 
-            if (!m_pExtract->Extract(m_hWnd, GetDlgItem(IDC_PROGRESS)))
+            if (m_hExtractionThread != INVALID_HANDLE_VALUE)
             {
-                /* Extraction failed, do not go to the next page */
+                CloseHandle(m_hExtractionThread);
```

Comment:
You should wait on the thread handle here in order to ensure it's actually finished executing (this can be important if the dll gets unloaded)

Reply:
For information, during the first try at modifying this, I was using WaitForSingleObject() to wait for the thread to stop in the case of a cancel, but it was freezing the whole program indefinitely, even though I couldn't find any deadlock in the code.

Reply:
Note: that last comment was just me not knowing how SendMessage (PropSheet_PressButton here) works. Replaced it with SendMessageCallback if that's acceptable (unless I manage to avoid re-entering OnWizardNext() ).